---
title: "Transposable elements"
author: "Moray Smith"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import libraries

```{r, warning = FALSE}
library(Biostrings)
library(tidyverse)
library(ggpubr)
library(ggthemes)
```

Read in the curated TE libraries for each annotation and calculate the mean size and count of classifications in each.

```{r}
read_te_library <- function(fasta, source) {
  library <- readDNAStringSet(fasta)
  family <- str_extract(names(library), "^(.*?)#(.*?)(?:/(.*))?$", group = 1)
  classification <- str_extract(names(library), "^(.*?)#(.*?)(?:/(.*))?$", group = 2)
  subfamily <- str_extract(names(library), "^(.*?)#(.*?)(?:/(.*))?$", group = 3)
  sizes <- width(library)
  df <- data.frame(family, classification, subfamily, sizes, source)
  return (df)
  }

earlgrey_library <- read_te_library(
  "../results/earlgrey/solanum_verrucosum-families.fa.strained",
  "earlgrey"
  )

edta_library <- read_te_library(
  "../results/edta/final_assembly.fa.mod.EDTA.TElib.fa",
  "edta"
  )

merged_library <- rbind(earlgrey_library, edta_library) |>
  mutate(classification = str_trim(classification)) |>
  mutate(
    classification = case_when(
      str_detect(classification, "Satellite|Simple_repeat|snRNA|tRNA|rRNA|MITE") ~ "other",
      str_detect(subfamily, "Helitron") ~ "helitron",
      .default = classification
      )
    )
```

Plot them

```{r}
library_size_plot <- ggplot(merged_library, aes(x = classification, y = sizes, fill = source)) +
  geom_boxplot(position = position_dodge(preserve = "single")) +
  scale_y_continuous(
    "Length (bp)",
    labels = scales::label_number(scale_cut = scales::cut_short_scale())
    ) +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau()

library_count_plot <- merged_library |>
  dplyr::count(source, classification) |>
  ggplot(aes(x = classification, y = n, fill = source)) +
  geom_bar(
    colour = "#333333",
    stat = "identity",
    position = position_dodge(preserve = "single")
    ) +
  ylab("Total #") +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau()
```
Now use gff files of annotations to calculate their total coverage across the genome.

```{r}
edta_gff <- "../results/edta/final_assembly.fa.mod.EDTA.TEanno.gff3" |>
  read_tsv(comment = "#", col_names = FALSE) |>
  mutate(
    family = str_extract(X9, "Name=(.*?);", group = 1),
    length = X5 - X4
    ) |>
  select(family, length)

earlgrey_gff <- "../results/earlgrey/solanum_verrucosum_summaryFiles/solanum_verrucosum.filteredRepeats.gff" |>
  read_tsv(comment = "#", col_names = FALSE) |>
  mutate(
    family = str_to_lower(str_extract(X9, "ID=(.*?);", group = 1)),
    length = X5 - X4
    ) |>
  select(family, length)

library_total_plot <- rbind(edta_gff, earlgrey_gff) |>
  left_join(merged_library, by = "family") |>
  group_by(classification, source) |>
  summarise(total_length = sum(length)) |>
  filter(!is.na(classification)) |> # handles 'repeat region EDTA'
  ggplot(aes(x = classification, y = total_length, fill = source)) +
  geom_bar(
    colour = "#333333",
    stat = "identity",
    position = position_dodge(preserve = "single")
    ) +
  scale_y_continuous(
    "Coverage (bp)",
    labels = scales::label_number(scale_cut = scales::cut_short_scale())
    ) +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau()
```
```{r}
ggsave(
  "plots/te_library_comparison.png",
  ggarrange(
    library_size_plot + rremove("xlab"),
    library_count_plot + rremove("xlab"),
    library_total_plot + rremove("xlab"),
    ncol = 1,
    align = "hv",
    common.legend = TRUE,
    labels = "auto",
    font.label = list(size = 10)
  ),
  width = 5.9,
  height = 4,
  dpi = 600
)
```

Import tesorter classifications of earlgrey
```{r}
tesorter <- read_tsv("../results/tesorter/earlgrey.cls.tsv") |>
  mutate(
    family = str_to_lower(
      str_extract(TE, "^(.*?)#(.*?)(?:/(.*))?$", group = 1)
      )
  )
```

```{r}
divergence_gff <- read_tsv(
  "../results/earlgrey/solanum_verrucosum_summaryFiles/solanum_verrucosum.filteredRepeats.divergence.gff",
  col_names = FALSE,
  ) |>
  filter(str_detect(X3, "LTR|LINE|SINE|DNA|Helitron|Unknown")) |>
  mutate(
    divergence = as.numeric(str_extract(X9, "KIMURA80=(.*?)$", group = 1)),
    length = X5 - X4,
    family = str_to_lower(str_extract(X9, "ID=(.*?);", group = 1))
    ) |>
  filter(!is.na(divergence) & divergence < 0.4) |>
  left_join(earlgrey_library, by = "family")
```

```{r}
kimura_full_plot <- divergence_gff |>
  mutate(
    classification = case_when(
      str_detect(X3, "LTR") ~ "LTR",
      str_detect(X3, "DNA") ~ "DNA",
      str_detect(X3, "LINE") ~ "LINE",
      str_detect(X3, "SINE") ~ "SINE",
      str_detect(X3, "Helitron") ~ "Helitron",
      str_detect(X3, "Unknown") ~ "Unknown"
      )
    ) |>
  ggplot(aes(x = divergence, weight = length, fill = classification)) +
  geom_histogram(bins = 50) +
  scale_y_continuous(
    "Coverage (bp)",
    labels = scales::label_number(scale_cut = scales::cut_short_scale())
    ) +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau(name = "TE classification")
kimura_full_plot
```

```{r}
kimura_ty3_plot <- divergence_gff |>
  left_join(tesorter, by = "family") |>
  filter(str_detect(Superfamily, "Gypsy"), Clade != "unknown") |>
  ggplot(aes(x = divergence, weight = length, fill = Clade)) +
  geom_histogram(bins = 50) +
  scale_y_continuous(
    "Coverage (bp)",
    labels = scales::label_number(scale_cut = scales::cut_short_scale())
    ) +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau(name = "Ty3 subclass")

kimura_ty3_plot
```

```{r}
kimura_ty1_plot <- divergence_gff |>
  left_join(tesorter, by = "family") |>
  filter(str_detect(Superfamily, "Copia"), Clade != "unknown") |>
  ggplot(aes(x = divergence, weight = length, fill = Clade)) +
  geom_histogram(bins = 50) +
  scale_y_continuous(
    "Coverage (bp)",
    labels = scales::label_number(scale_cut = scales::cut_short_scale())
    ) +
  theme_bw(base_size = 8) + theme(panel.grid = element_blank()) +
  scale_fill_tableau(name = "Ty1 subclass")

kimura_ty1_plot
```

```{r}
ggsave(
  "plots/kimura_density.png",
  ggarrange(
    kimura_full_plot + guides(fill = guide_legend(nrow = 5)),
    kimura_ty3_plot + guides(fill = guide_legend(nrow = 5)),
    kimura_ty1_plot + guides(fill = guide_legend(nrow = 5)),
    nrow = 3,
    align = "hv",
    labels = "auto",
    font.label = list(size = 10)
  ),
  width = 5.9,
  height = 5.9,
  dpi = 600,

)
```

Methylation profiles

```{r}
profiles <- read_csv("../results/te_methylation_summary.csv")

profile_plot <- ggplot(profiles) +
  geom_line(aes(x = Position, y = methylation, colour = type)) +
  facet_wrap(~ te_type, ncol = 5) +
  theme_bw(base_size = 8) +
  theme(panel.grid = element_blank()) +
  scale_colour_tableau(name = "Methylation") +
  scale_x_continuous(breaks = c(0, 200, 400, 600), labels = c("-2k", "5'", "3'", "2k")) +
  labs(x = "Position (bp)", y = "Mean methylation (prop.)")

cowplot::ggsave2(
  "plots/methylation_profile.png",
  profile_plot,
  width = 5.9,
  height = 5,
  dpi = 600
)
  
```

Save results
```{r}
earlgrey_library |>
  left_join(tesorter, by = "family")
```